doc/tapscript_opcodes.md

```diff
@@ -12,82 +12,72 @@ Taproot already increases a lot of resource limitations from segwitv0, so there
+      1. Define `OP_SUCCESS200` as `OP_INSPECTINPUTASSET`: Pop a `CScriptNum` input index `idx` and push the `nAsset` onto the stack as two elements. The first push the assetID(32), followed by the prefix(1)
+      1. Define `OP_SUCCESS201` as `OP_INSPECTINPUTVALUE`: Pop a `CScriptNum` input index `idx` and push the `nValue` as a tuple, value(8, 32) followed by prefix(1),
+      1. Define `OP_SUCCESS202` as `OP_INSPECTINPUTSCRIPTPUBKEY`: Pop a `CScriptNum` input index `idx` and push the following depending the type of scriptPubkey:
+         - If the scriptPubKey is not a native segwit program, fail if the size of scriptPubKey is more than 520(`MAX_SCRIPT_ELEMENT_SIZE`) bytes, otherwise push the scriptPubKey on stack top. Next, push a `CScriptNum(-1)` to indicate a non-native segwit scriptPubKey.
+         - If the scriptPubKey is a native segwit program, push the witness program(2-40) followed by a push for segwit version(0-1).
+      1. Define `OP_SUCCESS203` as `OP_INSPECTINPUTSEQUENCE`: Pop a `CScriptNum` input index `idx` and push the `nSequence`(4) as little-endian number.
+      1. Define `OP_SUCCESS204` as `OP_INSPECTINPUTISSUANCE`: Pop a `CScriptNum` input index `idx` and push the assetIssuance information if the asset has issuance, otherwise push an empty vector. Asset Issuance information is pushed as follows
+         - Push `nInflationKeys` as tuple, value(8, 32) followed by push for prefix(1). In case `nInflationKeys` is null, push an empty vector on stack top.
+         - Push `nAmount` as a tuple, value(8, 32) followed by a push for prefix(1). This cannot be null
+         - Push 32 byte `assetEntropy`
+         - Push 32 byte `assetBlindingNonce`
+   - Define `OP_SUCCESS205` as `OP_PUSHCURRENTINPUTINDEX` that pushes the current input index as `CScriptNum`. This can be used in conjunction with input introspection opcodes for inspecting current input.
+   - Output introspection opcodes:
+      1. Define `OP_SUCCESS206` as `OP_INSPECTOUTPUTASSET`: Pop a `CScriptNum` input index `idx` and push the `nAsset` as a tuple, first push the assetID(32), followed by the prefix(1)
+      1. Define `OP_SUCCESS207` as `OP_INSPECTOUTPUTVALUE`: Pop a `CScriptNum` input index `idx` and push the `nValue` as a tuple, value(8, 32) followed by prefix
+      1. Define `OP_SUCCESS208` as `OP_INSPECTOUTPUTNONCE`: Pop a `CScriptNum` input index `idx` and push the `nNonce`(33) onto the stack. If the nonce is null, push an empty vector onto the stack.
+      1. Define `OP_SUCCESS209` as `OP_INSPECTOUTPUTSCRIPTPUBKEY`: Pop a `CScriptNum` input index `idx` and push the scriptPubkey onto the stack.
```

Comment:
Pegouts and their PAK proofs won't fit in 520 bytes.  So we will need to push the hash of the scriptpubkey in the non-segwit case.  Then they can, in principle be verified with the rolling SHA256 opcodes.

However the implementation will require a cache to avoid DOS.